### DISCUSSION TITLE
getting rid of hollow 3d plot of the terrain

### DISCUSSION THREAD

**Discussion by aniketgupta2009** [pyvista/pyvista#2649] [Q&A]
I am sorry but I am really new to pyvista. I want to have a 3D plot of the terrain but I stuck with very unusual problem. My 3D plot boundaries are not closed and they look like a hollow. I have 11 depth layer and I individually checked them for their values, all layers looks good. However when I plot them in combine, I do not get closed boundaries. I am only getting the closed boundaries when the terrain touches the domain (i.e. where I do not have any nan values).

<img width="1280" alt="pyvista_plot" src="https://user-images.githubusercontent.com/50705068/130823724-8a0c4836-5ce4-45fc-8765-a7ced0629108.png">

import gemgis as gg
import numpy as np
import pyvista as pv
import matplotlib.pyplot as plt

file_name="try1.tif"
mesh1 = gg.visualization.read_raster(file_name,nodata_val=-99999,name='Elevation [m]')
topo = mesh1.warp_by_scalar()
z_cells = np.flip(np.array([10,10,10,10,10,10,10,10,10,10,10]))
xx = np.repeat(topo.x, len(z_cells), axis=-1)
yy = np.repeat(topo.y, len(z_cells), axis=-1)
zz = np.repeat(topo.z, len(z_cells), axis=-1) - np.cumsum(z_cells).reshape((1, 1, -1))
mesh = pv.StructuredGrid(xx, yy, zz)
mesh["Elevation"] = zz.ravel(order="F")
mesh.plot(show_edges=True)
plt.show()

Kindly help me !

Thank you in advance :)

**Comment by akaszynski** [comment]
Hello,
When you get a chance, could you upload the tif so I could take a look at this?

**Comment by aniketgupta2009** [comment]
Hi, thank you so much for your reply.

Here is the link to the file:
https://www.dropbox.com/s/qx44a4na726kqj0/try1.tif?dl=0

**Comment by RichardScottOZ** [comment]
So why the 11 layers?

**Comment by aniketgupta2009** [comment]
> So why the 11 layers?

These are heterogeneous underground layers to be represented in my figure. Usually they are not of the equal depth but for simplicity in above code I kept a constant length for these layer.

**Comment by akaszynski** [comment]
With #1620, you can:

```py
import gemgis as gg
import numpy as np
import pyvista as pv
import matplotlib.pyplot as plt

file_name = "try1.tif"
mesh1 = gg.visualization.read_raster(file_name,
                                     nodata_val=-99999,
                                     name='Elevation [m]')
topo = mesh1.warp_by_scalar()

z_cells = np.flip(np.array([10,10,10,10,10,10,10,10,10,10,10]))
xx = np.repeat(topo.x, len(z_cells), axis=-1)
yy = np.repeat(topo.y, len(z_cells), axis=-1)
zz = np.repeat(topo.z, len(z_cells), axis=-1) - np.cumsum(z_cells).reshape((1, 1, -1))


mesh = pv.StructuredGrid(xx, yy, zz)
mesh["Elevation"] = zz.ravel(order="F")

# must hide points
mesh.hide_points(np.isnan(mesh["Elevation"]))

mesh.plot(show_edges=True, scalars='Elevation')
```

![tmp](https://user-images.githubusercontent.com/11981631/131516233-1de10c68-da01-4d25-81db-8b21bf31c2dd.png)

**Comment by aniketgupta2009** [comment]
Hi, thank you so much for this. I can see that you succeded in plotting.

But, for me it still says an error.
**AttributeError: 'StructuredGrid' object has no attribute 'hide_points'**

I tried to "cast_to_unstructuredgrid", I tried creating the logical of ghost points (ghosts = np.argwhere(np.isnan(mesh["Elevation"]))  and then "mesh.remove_cells" along with several other options. But, nothing is working for me.

**Comment by RichardScottOZ** [comment]
I think Alex means coming in new pull request...you would have to install that version or add the new code.

**Comment by aniketgupta2009** [comment]
> I think Alex means coming in new pull request...you would have to install that version or add the new code.

Thank you so much ! Yes, I am able to do it now.

I was not able to install the new version, so I just took the edited function and appended it to my old file. It's working fine for me now. I will try to get the latest version if I am gonna to face any problem :-)

**Comment by akaszynski** [comment]
> I was not able to install the new version, so I just took the edited function and appended it to my old file. It's working fine for me now. I will try to get the latest version if I am gonna to face any problem :-)

That works.  If you want, you could have checked out the branch and run `pip install -e .` from the cloned repo, but waiting until the next release works too.

Closing this as it's resolved.
